libbpf-tools/tcppktlat.bpf.c

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2023 Wenbo Zhang
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+
+#include "compat.bpf.h"
+#include "tcppktlat.h"
+
+#define MAX_ENTRIES	10240
+#define AF_INET		2
+
+const volatile pid_t targ_pid = 0;
+const volatile pid_t targ_tid = 0;
+const volatile __u16 targ_sport = 0;
+const volatile __u16 targ_dport = 0;
+const volatile __u64 targ_min_us = 0;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, u64);
+} start SEC(".maps");
+
+static int handle_tcp_probe(struct sock *sk, struct sk_buff *skb)
+{
+	const struct inet_sock *inet = (struct inet_sock *)(sk);
+	u64 sock_cookie, ts, len, doff;
+	const struct tcphdr *th;
+
+	if (targ_sport && targ_sport != BPF_CORE_READ(inet, inet_sport))
+		return 0;
+	if (targ_dport && targ_dport != BPF_CORE_READ(sk, __sk_common.skc_dport))
+		return 0;
+	th = (const struct tcphdr*)BPF_CORE_READ(skb, data);
+	doff = BPF_CORE_READ_BITFIELD_PROBED(th, doff);
+	len = BPF_CORE_READ(skb, len);
+	// `doff * 4` means `__tcp_hdrlen`
+	if (len <= doff * 4)
+		return 0;
+	sock_cookie = bpf_get_socket_cookie(sk);
```

Comment:
`bpf_get_socket_cookie` does not work on old kernels.

Reply:
Yes, @yonghong-song 's advice is https://github.com/iovisor/bcc/pull/4574#discussion_r1202809133, any better idea?

Reply:
Could we just use `struct sock *sk` itself as key ?

Reply:
>  u64 bpf_get_socket_cookie(struct sk_buff *skb)

              Description
                     If the struct sk_buff pointed by skb has a known
                     socket, retrieve the cookie (generated by the
                     kernel) of this socket.  If no cookie has been set
                     yet, generate a new cookie. Once generated, the
                     socket cookie remains stable for the life of the
                     socket. This helper can be useful for monitoring
                     per socket networking traffic statistics as it
                     provides a global socket identifier that can be
                     assumed unique.

              Return A 8-byte long non-decreasing number on success, or
                     0 if the socket field is missing inside skb.

> The main purpose of generating cookies for Socket is to ensure the security and uniqueness of network connections, prevent data packets from being hijacked or forged, and at the same time enable the operating system to accurately know and handle different network connections.

I'm not sure if there are corner cases for using `struct sock *sk` directly.

Reply:
@iamkafai Would you please help with this? Thank you!

Reply:
Right, `struct sock *sk` as the argument is correct. The following is the kernel corresponding helper implementation.
```
BPF_CALL_1(bpf_get_socket_ptr_cookie, struct sock *, sk)
{       
        return sk ? sock_gen_cookie(sk) : 0;
}
```